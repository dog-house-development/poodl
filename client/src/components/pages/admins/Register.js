import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import AdminActions from '../../../actions/adminActions';
import _ from 'lodash';

import Form from '../../ui/Form';

const propTypes = {
    auth: PropTypes.object.isRequired,
    errors: PropTypes.object.isRequired
};

export class Register extends Component {
    constructor(props) {
        super(props);
        this.state = {
            firstName: '',
            lastName: '',
            email: '',
            password: '',
            password2: '',
            seniorCenterId: props.adminSeniorCenterId,
            superAdmin: false
        };
    }

    onChange = e => {
        this.setState({ [e.target.id]: e.target.value });
    };

    onSubmit = e => {
        e.preventDefault();

        const newAdmin = {
            firstName: this.state.firstName,
            lastName: this.state.lastName,
            email: this.state.email,
            password: this.state.password,
            password2: this.state.password2,
            seniorCenterId: this.state.seniorCenterId,
            superAdmin: this.state.superAdmin
        };

        this.props.adminActions.create(newAdmin, this.props.history);
    };

    getFields = () => {
        const { errors } = this.props;
        const fields = [
            {
                onChange: this.onChange,
                error: errors.firstName,
                id: 'firstName',
                type: 'text',
                label: 'First name',
                placeholder: 'John...',
                sidebyside: 1,
                autoComplete: 'off'
            },
            {
                onChange: this.onChange,
                error: errors.lastName,
                id: 'lastName',
                type: 'text',
                label: 'Last name',
                placeholder: 'Smith...',
                sidebyside: 2,
                autoComplete: 'off'
            },
            {
                onChange: this.onChange,
                error: errors.email,
                id: 'email',
                type: 'email',
                label: 'Email',
                placeholder: 'example@poodl.com...',
                autoComplete: 'off'
            },
            {
                onChange: this.onChange,
                error: errors.password,
                id: 'password',
                type: 'password',
                label: 'Password',
                placeholder: 'Shhhhh...',
                autoComplete: 'off'
            },
            {
                onChange: this.onChange,
                error: errors.password2,
                id: 'password2',
                type: 'password',
                label: 'Confirm Password',
                placeholder: 'Again...',
                autoComplete: 'off'
            }
        ];
        if (this.props.adminIsSuper) {
            fields.push({
                onChange: this.onChange,
                content: this.props.adminSeniorCenterId,
                error: errors.seniorCenterId,
                id: 'seniorCenterId',
                type: 'text',
                label: 'Senior Center Id',
                placeholder: 'ID...'
            });
        }
        return fields;
    };

    render() {
        return (
            <div className="register-container">
                <Link to="/dashboard" className="button small tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to home
                </Link>
                <Form
                    noValidate
                    fields={this.getFields()}
                    onSubmit={this.onSubmit}
                    buttonLabel="Register"
                    formTitle="Register Admin"
                />
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        auth: state.auth,
        adminIsSuper: _.get(state.auth.admin, 'accessLevel', false) === 'Super',
        adminSeniorCenterId: _.get(state.auth.admin, 'seniorCenterId'),
        errors: state.admins.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        adminActions: bindActionCreators(AdminActions, dispatch)
    };
};

Register.propTypes = propTypes;
export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(Register));
