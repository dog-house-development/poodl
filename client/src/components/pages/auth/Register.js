import React, { Component } from 'react';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { registerAdmin } from '../../../actions/authActions';

import GoHome from '../../ui/GoHome';
import Form from '../../ui/Form';

const propTypes = {
    registerAdmin: PropTypes.func.isRequired,
    auth: PropTypes.object.isRequired,
    errors: PropTypes.object.isRequired
};

export class Register extends Component {
    constructor() {
        super();
        this.state = {
            name: '',
            email: '',
            password: '',
            password2: '',
            errors: {}
        };
    }

    componentDidMount() {
        // If logged in and admin navigates to Register page, should redirect them to dashboard
        if (this.props.auth.isAuthenticated) {
            this.props.history.push('/dashboard');
        }
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.errors) {
            this.setState({
                errors: nextProps.errors
            });
        }
    }

    onChange = e => {
        this.setState({ [e.target.id]: e.target.value });
    };

    onSubmit = e => {
        e.preventDefault();

        const newAdmin = {
            name: this.state.name,
            email: this.state.email,
            password: this.state.password,
            password2: this.state.password2
        };

        this.props.registerAdmin(newAdmin, this.props.history);
    };

    getFields = () => {
        const { errors } = this.state;
        const fields = [
            {
                onChange: this.onChange,
                value: this.state.name,
                error: errors.name,
                id: 'name',
                type: 'text',
                label: 'Name',
                placeholder: 'First Last...'
            },
            {
                onChange: this.onChange,
                value: this.state.email,
                error: errors.email,
                id: 'email',
                type: 'email',
                label: 'Email',
                placeholder: 'example@poodl.com...'
            },
            {
                onChange: this.onChange,
                value: this.state.password,
                error: errors.password,
                id: 'password',
                type: 'password',
                label: 'Password',
                placeholder: 'Shhhhh...'
            },
            {
                onChange: this.onChange,
                value: this.state.password2,
                error: errors.password2,
                id: 'password2',
                type: 'password',
                label: 'Confirm Password',
                placeholder: 'Again...'
            }
        ];
        return fields;
    };

    render() {
        return (
            <div className="register-container">
                <GoHome to="/" className="button small tertiary" buttonOrLink="button" size="small" kind="tertiary">
                    <i className="material-icons">keyboard_backspace</i> Back to home
                </GoHome>
                <Form
                    noValidate
                    fields={this.getFields()}
                    onSubmit={this.onSubmit}
                    buttonLabel="Register"
                    formTitle="Register Admin"
                />
                <div className="center">
                    <p>
                        Already have an account?{' '}
                        <Link to="/login" className="button small secondary">
                            Log in
                        </Link>
                    </p>
                </div>
            </div>
        );
    }
}

export const mapStateToProps = (state, props) => {
    return {
        auth: state.auth,
        errors: state.errors
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        registerAdmin: (adminData, history) => dispatch(registerAdmin(adminData, history))
    };
};

Register.propTypes = propTypes;
export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(Register));
