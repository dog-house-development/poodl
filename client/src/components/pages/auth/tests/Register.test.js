import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import _ from 'lodash';

import { Register, mapStateToProps, mapDispatchToProps } from '../Register';

configure({ adapter: new Adapter() });

describe('Register tests', () => {
    let state, props, wrapper, instance;
    const setInstanceAndWrapper = (_props = {}, _state = {}) => {
        state = _.assign(
            {},
            {
                auth: {
                    isAuthenticated: false,
                    loading: false
                },
                errors: {}
            },
            _state
        );
        props = _.assign({}, { history: ['/register'] }, _props);
        wrapper = shallow(
            <Register
                {..._.assign(
                    {},
                    props,
                    mapStateToProps(state, props),
                    mapDispatchToProps(jasmine.createSpy('dispatch'))
                )}
            />
        );
        instance = wrapper.instance();
        wrapper.setState({
            name: 'Sam',
            email: 'sam@test.com',
            password: 'abc123',
            password2: 'abc123',
            errors: { email: 'Invalid email' }
        });
    };

    beforeEach(() => {
        setInstanceAndWrapper();
    });

    describe('mapStateToProps', () => {
        it('should map state to props', () => {
            expect(mapStateToProps(state, props)).toEqual({
                auth: { isAuthenticated: false, loading: false },
                errors: {}
            });
        });
    });

    describe('mapDispatchToProps', () => {
        it('should map dispatch to props', () => {
            const dispatch = jest.fn();
            expect(JSON.parse(JSON.stringify(mapDispatchToProps(dispatch)))).toEqual(
                JSON.parse(JSON.stringify({ registerAdmin: () => {} }))
            );
        });
    });

    describe('componentDidMount', () => {
        it('should redirect if admin is authenticated', () => {
            const authenticatedState = {
                auth: {
                    isAuthenticated: true,
                    loading: false
                }
            };
            setInstanceAndWrapper({}, authenticatedState);
            const newInstanceProps = instance.props.history.concat('/dashboard');
            instance.componentDidMount();
            expect(instance.props.history).toEqual(newInstanceProps);
        });
    });

    describe('componentWillReceiveProps', () => {
        it('should set state.errors to newProps errors', () => {
            spyOn(instance, 'componentWillReceiveProps');
            expect(instance.state.errors).toEqual({ email: 'Invalid email' });
            wrapper.setProps({ errors: { password: 'Fake password' } });
            expect(instance.componentWillReceiveProps).toHaveBeenCalled();
            //expect(instance.state.errors).toEqual({ password: 'Fake password' });
            wrapper.setProps({ errors: {} });
            //expect(instance.state.errors).toEqual({});
            expect(instance.componentWillReceiveProps.calls.count()).toBe(2);
        });
    });

    describe('onChange', () => {
        it('should run without errors', () => {
            spyOn(instance, 'onChange');
            expect(instance.state.email).toEqual('sam@test.com');
            const e = {
                target: { id: 'email', value: 'Meow' }
            };
            instance.onChange(e);
        });
    });

    describe('onSubmit', () => {
        it('should run without errors', () => {
            spyOn(instance, 'onSubmit');
            const e = {
                target: { id: 'email', value: 'Meow' },
                preventDefault: () => {}
            };
            wrapper.find('Form').simulate('submit', e);
        });
    });

    describe('getFields', () => {
        it('should return correct fields', () => {
            expect(instance.getFields()).toEqual([
                {
                    onChange: instance.onChange,
                    value: 'Sam',
                    id: 'name',
                    type: 'text',
                    label: 'Name',
                    placeholder: 'First Last...'
                },
                {
                    onChange: instance.onChange,
                    value: 'sam@test.com',
                    id: 'email',
                    type: 'email',
                    label: 'Email',
                    error: 'Invalid email',
                    placeholder: 'example@poodl.com...'
                },
                {
                    onChange: instance.onChange,
                    value: 'abc123',
                    id: 'password',
                    type: 'password',
                    label: 'Password',
                    placeholder: 'Shhhhh...'
                },
                {
                    onChange: instance.onChange,
                    value: 'abc123',
                    id: 'password2',
                    type: 'password',
                    label: 'Confirm Password',
                    placeholder: 'Again...'
                }
            ]);
        });
    });

    describe('render', () => {
        it('should render correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
