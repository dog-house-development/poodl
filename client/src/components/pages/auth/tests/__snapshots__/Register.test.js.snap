// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register tests render should render correctly 1`] = `
<div
  className="register-container"
>
  <Connect(GoHome)
    buttonOrLink="button"
    className="button small tertiary"
    kind="tertiary"
    size="small"
    to="/"
  >
    <i
      className="material-icons"
    >
      keyboard_backspace
    </i>
     Back to home
  </Connect(GoHome)>
  <Form
    buttonLabel="Register"
    fields={
      Array [
        Object {
          "error": undefined,
          "id": "name",
          "label": "Name",
          "onChange": [Function],
          "placeholder": "First Last...",
          "type": "text",
          "value": "Sam",
        },
        Object {
          "error": "Invalid email",
          "id": "email",
          "label": "Email",
          "onChange": [Function],
          "placeholder": "example@poodl.com...",
          "type": "email",
          "value": "sam@test.com",
        },
        Object {
          "error": undefined,
          "id": "password",
          "label": "Password",
          "onChange": [Function],
          "placeholder": "Shhhhh...",
          "type": "password",
          "value": "abc123",
        },
        Object {
          "error": undefined,
          "id": "password2",
          "label": "Confirm Password",
          "onChange": [Function],
          "placeholder": "Again...",
          "type": "password",
          "value": "abc123",
        },
      ]
    }
    formTitle="Register Admin"
    noValidate={true}
    onSubmit={[Function]}
  />
  <div
    className="center"
  >
    <p>
      Already have an account?
       
      <Link
        className="button small secondary"
        replace={false}
        to="/login"
      >
        Log in
      </Link>
    </p>
  </div>
</div>
`;
